src/rgw/rgw_putobj_processor.cc

```diff
@@ -500,4 +500,168 @@ int MultipartObjectProcessor::complete(size_t accounted_size,
   return 0;
 }
 
+int AppendObjectProcessor::process_first_chunk(bufferlist &&data, rgw::putobj::DataProcessor **processor)
+{
+  int r = writer.write_exclusive(data);
+  if (r < 0) {
+    return r;
+  }
+  *processor = &stripe;
+  return 0;
+}
+
+int AppendObjectProcessor::prepare()
+{
+  RGWObjState *astate;
+  int r = store->get_obj_state(&obj_ctx, bucket_info, head_obj, &astate);
+  if (r < 0) {
+    return r;
+  }
+  cur_size = astate->size;
+  *cur_accounted_size = astate->accounted_size;
+  if (!astate->exists) {
+    if (position != 0) {
+      ldout(store->ctx(), 5) << "ERROR: Append position should be zero" << dendl;
+      return -ERR_POSITION_NOT_EQUAL_TO_LENGTH;
+    } else {
+      cur_part_num = 1;
+      //set the prefix
+      char buf[33];
+      gen_rand_alphanumeric(store->ctx(), buf, sizeof(buf) - 1);
+      string oid_prefix = head_obj.key.name;
+      oid_prefix.append(".");
+      oid_prefix.append(buf);
+      oid_prefix.append("_");
+      manifest.set_prefix(oid_prefix);
+    }
+  } else {
+    // check whether the object appendable
+    map<string, bufferlist>::iterator iter = astate->attrset.find(RGW_ATTR_APPEND_PART_NUM);
+    if (iter == astate->attrset.end()) {
+      ldout(store->ctx(), 5) << "ERROR: The object is not appendable" << dendl;
+      return -ERR_OBJECT_NOT_APPENDABLE;
+    }
+    if (position != *cur_accounted_size) {
+      ldout(store->ctx(), 5) << "ERROR: Append position should be equal to the obj size" << dendl;
+      return -ERR_POSITION_NOT_EQUAL_TO_LENGTH;
+    }
+    try {
+      decode(cur_part_num, iter->second);
+    } catch (buffer::error& err) {
+      ldout(store->ctx(), 5) << "ERROR: failed to decode part num" << dendl;
+      return -EIO;
+    }
+    cur_part_num++;
+    //get the current obj etag
+    iter = astate->attrset.find(RGW_ATTR_ETAG);
+    if (iter != astate->attrset.end()) {
+      string s = rgw_string_unquote(iter->second.c_str());
+      size_t pos = s.find("-");
+      cur_etag = s.substr(0, pos);
+    }
+    cur_manifest = &astate->manifest;
```

Comment:
If an appendobject request happens to change astate-manifest after this thread read astate, then the cur_manifest here will be obsolete. Based on this obsolete manifest, the data of the prevous appendobject will be lost. So appendobject seems no quite safe.

Reply:
Yes, append data to the same object concurrently is not safe.

Reply:
@Jeegn-Chen are you suggesting that two append ops could race to access the same `RGWObjState::manifest` memory? that shouldn't be possible, because each rgw op uses a separate instance of `RGWObjectCtx` from `req_state::obj_ctx`

or are you saying that ops could race to overwrite RGW_ATTR_MANIFEST on the head object? i believe we covered this case with the logic i described here: https://github.com/ceph/ceph/pull/22755#discussion_r230492710

> and two racing appends to an existing object will use the same prefix and part number, so the write_exclusive() in process_first_chunk() will only succeed for the first writer